### PR TITLE
Fix warnings when files are overwritten in the build directory.

### PR DESCRIPTION
**Pull request by picnixz**
cc @AA-Turner 

I fixed the test (it was due to the two files having the same file size, and `filecmp.cmp()` returned True because it did not checked the contents). I also added a message for the assertion since it helps in debugging the error location.

### CODE REVIEW

**Comment by AA-Turner** [comment]
I don't want to have a new warning in a micro release, hence using `info` - please revert.

A

File: sphinx/util/osutil.py [hunk 1]
```diff
@@ -106,7 +106,12 @@ def copyfile(
         msg = f'{os.fsdecode(source)} does not exist'
         raise FileNotFoundError(msg)
 
-    if not (dest_exists := path.exists(dest)) or not filecmp.cmp(source, dest):
+    if (
+        not (dest_exists := path.exists(dest)) or
+        # comparison must be done using shallow=False since
+        # two different files might have the same bytes size
```
Comment (AA-Turner) [inline sphinx/util/osutil.py]:
```suggestion
        # two different files might have the same size
```

**Comment by picnixz** [comment]
> I don't want to have a new warning in a micro release, hence using info - please revert.

I updated the tests accordingly (since the message would no more be printed on the warnings stream).

File: sphinx/util/fileutil.py [hunk 2]
```diff
@@ -75,6 +75,8 @@ def copy_asset_file(source: str | os.PathLike[str], destination: str | os.PathLi
             msg = ('Copying the rendered template %s to %s will overwrite data, '
                    'as a file already exists at the destination path '
                    'and the content does not match.')
+            # See https://github.com/sphinx-doc/sphinx/pull/12627#issuecomment-2241144330
+            # for the rational for logger.info().
```
Comment (AA-Turner) [inline sphinx/util/fileutil.py]: SPaG:
```suggestion
            # for the rationale for logger.info().
```

Comment (picnixz) [reply sphinx/util/fileutil.py]: I committed too fast :') Thanks for spotting it!

**Comment by AA-Turner** [comment]
Thanks!

A

**Pull request merged**